### PR TITLE
#162414161 validate user's input data

### PR DESCRIPTION
**Pull request by Luckzman**
#### What does this PR do?
This PR validates users' input during signup
#### Description of Task to be completed?
It ensures that a user does enter invalid and blank input to our database, and if they try to input say a blank field, they will get a descriptive error message.
#### How should this be manually tested?
This can be manually tested by pulling the branch and running `db:migrate`

- git pull the current repository
- create databases for user 'User' 
- Switch to ft-signup-validation-#23162414161 branch
- Run `npm install` to install all dependencies.
- Run `db:migrate`
- To test the models using mocha run `npm test`

#### What are the relevant pivotal tracker stories?
PT Link: https://www.pivotaltracker.com/story/show/162414160
#### Screenshots
screenshot 2019-01-18 at 11 54 45 pm

### CODE REVIEW

File: controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/** Class to add new user and login registered user. */
+class AuthController {
```
Comment (codrex) [inline controllers/authController.js]: Why do we have this empty class?

File: database/config.js [hunk 2]
```diff
@@ -6,14 +6,14 @@ module.exports = {
     password: process.env.DATABASE_PASSWORD,
     database: process.env.DATABASE_NAME,
     host: process.env.DATABASE_HOST,
-    dialect: process.env.DATABASE_DIALECT
+    dialect: 'postgres'
```
Comment (codrex) [inline database/config.js]: Why this change?

File: lib/modelManagers/usermodel.js [hunk 3]
```diff
@@ -22,6 +22,22 @@ class User {
...
   }
 
+  /**
+   *
+   * @param {field} field The field to be searched.
+   * @static
+   * @memberof User
+   * @returns {object} The parameters from the query
+   */
+  static async findEmail(field) {
```
Comment (codrex) [inline lib/modelManagers/usermodel.js]: The name `findName` is kind of misleading, a better name would be `findUserByEmail`. The parameter `field` can be better written as `email`

File: middlewares/validators/authValidator.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable newline-per-chained-call */
+// const emailExist = require('../../lib/utils/checkers/emailExistChecker');
+// const userModelManager = require('../../lib/modelManagers/usermodel');
```
Comment (codrex) [inline middlewares/validators/authValidator.js]: Remove dead code

File: middlewares/validators/authValidator.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    .matches('[a-z]').withMessage('password must contain a lowercase')
+    .matches('[A-Z]').withMessage('password must contain a uppercase')
+    .matches('[0-9]').withMessage('password must contain a number');
+  req.check('firstname')
+    .notEmpty().withMessage('firstname is required')
+    .isAlpha().trim().withMessage('firstname must be alphabet letters');
+  req.check('lastname')
+    .notEmpty().withMessage('lastname is required')
+    .isAlpha().trim().withMessage('lastname must be alphabet letters');
+
```
Comment (codrex) [inline middlewares/validators/authValidator.js]: `req.check('firstname')
    .notEmpty().withMessage('firstname is required')
    .isAlpha().trim().withMessage('firstname must be alphabet letters');` is similar to 

`
req.check('lastname')
    .notEmpty().withMessage('lastname is required')
    .isAlpha().trim().withMessage('lastname must be alphabet letters');
`
consider abstracting it into a function. The function should look like the function below



```js

   function validateName(req, name) {
         req.check(name')
             .notEmpty().withMessage(`${name} is required`)
             .isAlpha().trim().withMessage(`${name} must be alphabet letters`);
       }  
```

**Review by Maxfurry — DISMISSED** [review]

File: middlewares/validators/authValidator.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  const errors = req.validationErrors();
+  if (errors) { return errorMessage(res, errors[0].msg); }
```
Comment (codrex) [inline middlewares/validators/authValidator.js]: Can you please provide context on what is going on here
`if (errors) { return errorMessage(res, errors[0].msg); }`

File: middlewares/validators/authValidator.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+});
+
+/**
+   * @description check against errors on signup fields
+   * @param {object} req
+   * @param {object} res
+   * @param {object} next
+   * @returns {object} returns error object if error exist
+   */
+const authValidator = (req, res, next) => {
```
Comment (codrex) [inline middlewares/validators/authValidator.js]: please rename this function to a more meaningful name, `signupValidator` is more preferable.

File: middlewares/validators/authValidator.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+const errorMessage = (res, message) => res.status(400).json({
+  status: false,
+  message
+});
+
+/**
+   * @description check against errors on signup fields
+   * @param {object} req
+   * @param {object} res
+   * @param {object} next
```
Comment (codrex) [inline middlewares/validators/authValidator.js]: `next` is a function and not an object

https://expressjs.com/en/guide/writing-middleware.html

File: middlewares/validators/authValidator.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+   * @returns {object} returns error object if error exist
+   */
+const authValidator = (req, res, next) => {
+  req.check('email')
+    .notEmpty().withMessage('Email is required')
+    .isEmail().trim().withMessage('Email must be valid');
+  req.check('password')
```
Comment (codrex) [inline middlewares/validators/authValidator.js]: Abstracting this into a function make the code cleaner and more reusable.

Please refer to the code snippet below for direction. 👇

File: middlewares/validators/authValidator.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+   * @description check against errors on signup fields
+   * @param {object} req
+   * @param {object} res
+   * @param {object} next
+   * @returns {object} returns error object if error exist
+   */
+const authValidator = (req, res, next) => {
+  req.check('email')
```
Comment (codrex) [inline middlewares/validators/authValidator.js]: Abstracting this into a function make the code cleaner and more reusable.

File: middlewares/validators/checkEmail.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ * @description Check if email is already in the database
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next middleware in the stack
+ * @returns {object} res object
+ */
+async function checkEmail(req, res, next) {
+  try {
+    const record = await User.findEmail(req.body.email);
+    if (record.dataValues.email) {
```
Comment (codrex) [inline middlewares/validators/checkEmail.js]: Doing this is dangerous. In a scenario where no record is found, null will be returned. performing a property lookup on `null` will throw an error. A better approach should be 

`if (record){} `

File: middlewares/validators/checkEmail.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      return res.status(404).json({
+        status: false,
+        message: 'Authentication Failed',
+      });
+    }
+  } catch (error) {
+    return next();
```
Comment (codrex) [inline middlewares/validators/checkEmail.js]: Why are you calling next in the catch block?

File: middlewares/validators/checkEmail.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+const User = require('../../lib/modelManagers/usermodel');
+
+/**
+ * @description Check if email is already in the database
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next middleware in the stack
+ * @returns {object} res object
+ */
+async function checkEmail(req, res, next) {
```
Comment (codrex) [inline middlewares/validators/checkEmail.js]: Please use a function name that best describes what the function is doing.

File: middlewares/validators/checkEmail.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  try {
+    const record = await User.findEmail(req.body.email);
+    if (record.dataValues.email) {
+      return res.status(404).json({
+        status: false,
+        message: 'Authentication Failed',
```
Comment (codrex) [inline middlewares/validators/checkEmail.js]: `'Authentication Failed'` is not a descriptive error message. A proper error message should be able to tell the user what caused the failure and ways of fixing the error if an action is required from the user to fix the error.

File: routes/api/auth.js [hunk 15]
```diff
@@ -0,0 +1,7 @@
+const router = require('express').Router();
+const authValidator = require('../../middlewares/validators/authValidator');
+const checkEmail = require('../../middlewares/validators/checkEmail');
+
+router.post('/auth/signup', authValidator, checkEmail);
```
Comment (codrex) [inline routes/api/auth.js]: I feel having `'/auth/signup'` as `'/signup'` is a better approach

File: routes/api/index.js [hunk 16]
```diff
@@ -1,16 +1,5 @@
...
-      errors: Object.keys(err.errors).reduce((errors, key) => {
-        errors[key] = err.errors[key].message;
-        return errors;
-      }, {})
-    });
-  }
-
-  return next(err);
-});
+router.post('/auth', require('./auth'));
```
Comment (codrex) [inline routes/api/index.js]: This `/auth` should be `/users`

**Review by codrex — CHANGES_REQUESTED** [review]

File: controllers/authController.js (hunk 1 above)
Comment (Luckzman) [reply controllers/authController.js]: fixed

File: database/config.js (hunk 2 above)
Comment (Luckzman) [reply database/config.js]: fixed

File: middlewares/validators/authValidator.js (hunk 5 above)
Comment (Luckzman) [reply middlewares/validators/authValidator.js]: fixed

File: middlewares/validators/checkEmail.js (hunk 12 above)
Comment (Luckzman) [reply middlewares/validators/checkEmail.js]: fixed

File: middlewares/validators/checkEmail.js (hunk 14 above)
Comment (Luckzman) [reply middlewares/validators/checkEmail.js]: fixed

File: middlewares/validators/authValidator.js (hunk 8 above)
Comment (Luckzman) [reply middlewares/validators/authValidator.js]: thanks

File: middlewares/validators/authValidator.js (hunk 9 above)
Comment (Luckzman) [reply middlewares/validators/authValidator.js]: done

File: middlewares/validators/authValidator.js (hunk 10 above)
Comment (Luckzman) [reply middlewares/validators/authValidator.js]: done

File: middlewares/validators/checkEmail.js (hunk 11 above)
Comment (Luckzman) [reply middlewares/validators/checkEmail.js]: noted, thanks a lot...I never thought of it that way

File: middlewares/validators/checkEmail.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next middleware in the stack
+ * @returns {object} res object
+ */
+async function checkEmail(req, res, next) {
+  try {
+    const record = await User.findUserByEmail(req.body.email);
+    if (record) {
+      return res.status(409).json({
```
Comment (codrex) [inline middlewares/validators/checkEmail.js]: move constants and string litral to the constants file

File: middlewares/validators/signupValidator.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable newline-per-chained-call */
+const { validateName, validateEmail, validatePassword } = require('../../lib/utils/validate');
+
+const errorMessage = (res, messages) => res.status(400).json({
```
Comment (codrex) [inline middlewares/validators/signupValidator.js]: Move this function into utils. Doing this will make the function reusable across application. Function names make more sense when it is a verb and not a noun.

File: middlewares/validators/signupValidator.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  status: false,
+  messages
+});
+
+/**
+   * @description check against errors on signup fields
+   * @param {object} req
+   * @param {object} res
+   * @param {function} next
+   * @returns {object} returns error object if error exist
```
Comment (codrex) [inline middlewares/validators/signupValidator.js]: This function returns undefined at all times

File: test/checkEmail.test.js [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+const User = require('../lib/modelManagers/usermodel');
+
+chai.use(sinonChai);
+const { expect, should } = chai;
+should();
+
+describe('check if email exist', () => {
+  afterEach(sinon.restore);
+
+  it('should find already existing user email and return conflict', async () => {
```
Comment (codrex) [inline test/checkEmail.test.js]: Your case description does not communicate what the test is doing. Reading the description for a test should fully describe the test. Please make your test descriptions a lot more descriptive.

**Review by codrex — COMMENTED** [review]
Nice stuff.

Attend to the comments

File: test/checkEmail.test.js (hunk 20 above)
Comment (Luckzman) [reply test/checkEmail.test.js]: fixed

File: middlewares/validators/checkEmail.js (hunk 17 above)
Comment (Luckzman) [reply middlewares/validators/checkEmail.js]: fixed

File: middlewares/validators/signupValidator.js (hunk 18 above)
Comment (Luckzman) [reply middlewares/validators/signupValidator.js]: done

File: middlewares/validators/signupValidator.js (hunk 19 above)
Comment (Luckzman) [reply middlewares/validators/signupValidator.js]: fixed

File: lib/utils/signupValidator.js [hunk 21]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable newline-per-chained-call */
+const { validateName, validateEmail, validatePassword } = require('./validate');
+const { failureResponse } = require('../utils/messageHandler');
+const { BAD_REQUEST_CODE } = require('../../constants/index');
+
+// const BAD_REQUEST_CODE = 400;
```
Comment (codrex) [inline lib/utils/signupValidator.js]: Remove dead code

File: lib/utils/signupValidator.js [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+
+const formatErrorMessage = (msgArr) => {
+  const message = {};
+  // eslint-disable-next-line no-restricted-syntax
+  for (const msg of msgArr) {
+    if (!message[msg.split(' ')[0]]) {
```
Comment (codrex) [inline lib/utils/signupValidator.js]: `message[msg.split(' ')`
It will be better to perform this operation once, save the result in a variable and reuse in other parts of the code that needs it.

File: lib/utils/signupValidator.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable newline-per-chained-call */
+const { validateName, validateEmail, validatePassword } = require('./validate');
+const { failureResponse } = require('../utils/messageHandler');
+const { BAD_REQUEST_CODE } = require('../../constants/index');
+
+// const BAD_REQUEST_CODE = 400;
+
+const formatErrorMessage = (msgArr) => {
```
Comment (codrex) [inline lib/utils/signupValidator.js]: move `formatErrorMessage ` to utils

File: middlewares/checkEmail.js [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next middleware in the stack
+ * @returns {*} *
+ */
+async function checkEmail(req, res, next) {
+  try {
+    const record = await User.findUserByEmail(req.body.email);
+    if (record) {
+      return failureResponse(res, 'email already exist', CONFLICT_CODE);
```
Comment (codrex) [inline middlewares/checkEmail.js]: move your string literal into  constants

File: test/checkEmail.test.js [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+
+chai.use(sinonChai);
+const { expect, should } = chai;
+should();
+describe('check email', () => {
+  describe('findUserByEmail Model Manager Test', async () => {
+    afterEach(sinon.restore);
+    it('should find user if email exist', async () => {
+      const email = 'lucky@gmail.com';
+      sinon.stub(user, 'findOne').returns(true);
```
Comment (codrex) [inline test/checkEmail.test.js]: Mocking out `findOne` make this test equivalent to testing if true is equal to true. Please reach out to @eltNEG  for directions on how to test model managers.

File: test/checkEmail.test.js [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+  checkEmail = require('../middlewares/checkEmail'),
+  User = require('../lib/modelManagers/usermodel'),
+  user = require('../database/models').User,
+  { SERVER_ERROR_CODE } = require('../constants');
+
+chai.use(sinonChai);
+const { expect, should } = chai;
+should();
+describe('check email', () => {
+  describe('findUserByEmail Model Manager Test', async () => {
```
Comment (codrex) [inline test/checkEmail.test.js]: Move model managers test into its own file

File: test/signupValidator.test.js [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+      .request(app)
+      .post('/api/v1/user/signup')
+      .send({
+        email: '',
+        password: 'testTEST@123',
+        firstname: 'firstname',
+        lastname: 'lastname'
+      })
+      .end((err, res) => {
+        res.should.have.status(BAD_REQUEST_CODE);
```
Comment (codrex) [inline test/signupValidator.test.js]: You should also assert for the message sent back from the server.

File: lib/utils/validate.js [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+    .withMessage(`${name} must be alphabets`);
+}
+/**
+ *
+ * @description This function validate input field that need an alphabet
+ * @param {object} req
+ * @param {string} email
+ * @returns {*} *
+ */
+function validateEmail(req, email) {
```
Comment (codrex) [inline lib/utils/validate.js:25]: You don't need to pass in `email` as an argument to the function since the function only validates email.

File: lib/utils/signupValidator.js [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+  const message = {};
+  // eslint-disable-next-line no-restricted-syntax
```
Comment (codrex) [inline lib/utils/signupValidator.js]: `// eslint-disable-next-line no-restricted-syntax`
Why are you disabling this rule?

File: test/emailVerification/sendVerificationMail.test.js [hunk 30]
```diff
@@ -3,32 +3,41 @@ const { expect } = require('chai');
...
+
+    const msg = {
+      to: 'michael.i.owolabi@gmail.com',
+      from: FROM,
+      subject: SUBJECT,
+      html: emailTemplate(hostUrl)
+    };
+    sinon.stub(sgMail, 'send').returns(undefined);
     sinon.stub(res, 'status').returnsThis();
+    // sinon.stub(sendVerificationMail, 'sendVerificationMail');
```
Comment (codrex) [inline test/emailVerification/sendVerificationMail.test.js]: remove dead code

**Review by codrex — APPROVED** [review]
